include/model/CModelFactory.h

```diff
@@ -417,6 +427,12 @@ class MODEL_EXPORT CModelFactory {
     //! The global model configuration parameters.
     SModelParams m_ModelParams;
 
+    //! A reference to the singleton interim bucket correction calculator.
+    //!
+    //! \note That this is stored by weak pointer since we don't want this
+    //! to update the reference count or our memory accounting will be wrong.
```

Comment:
I have bad feelings about the usage of weak pointers here, just for the purpose of workarounding memory accounting problems. I know this is a difficult problem but we should rather work on the root cause, than introducing such a workaround. (Although I guess that for this case its benign, I guess the logic ensures that the pointer never gets invalid before this class gets destructed)

Reply:
The ownership model of this object is actually very simple. The model config object, which really "owns it", is created early in main and is never deleted. So there is no chance that this goes away. All references to this could just as well be bare pointers, including those in the models. However, the reference counting is a convenient mechanism for dividing up the memory of the interim bucket corrector among the different models. This means in model debug we get sensible attribution of the memory to each individual model.

We don't have much precedence for globally owned state like this. One example is the model params which is owned by the model factories which are themselves owned by the model config and held by reference in the individual models. So equally dangerous if this object was deleted. I could equally hold this by reference to the single instance of the shared pointer here, but I don't feel that is an improvement. I thought about different ownership strategies, but didn't come up with a better one.

I guess as a minimum I should document this a better in the code, including the assumption of any code which uses the factories that they own the interim bucket corrector. But definitely willing to consider alternative suggestions.

Reply:
An alternative, I'm not sure I prefer it, would be to make something like a non-owning counted reference object. We understand that such an object doesn't imply shared ownership, so doesn't interfere with the referred to object lifecycle, but it does maintain a count of references which have been created. We'd need a weak version of this as well with similar semantics, but which doesn't throw when upgrading to a full reference. (This is awfully like a shared_ptr/weak_ptr, so I'm not sure I want to go to the effort of writing this class.)